program_transformers/src/mpl_core_program/v1_asset.rs

```diff
@@ -0,0 +1,558 @@
+    },
+    digital_asset_types::{
+        dao::{
+            asset, asset_authority, asset_creators, asset_data, asset_grouping,
+            sea_orm_active_enums::{
+                ChainMutability, Mutability, OwnerType, SpecificationAssetClass,
+            },
+        },
+        json::ChainDataV1,
+    },
+    heck::ToSnakeCase,
+    sea_orm::{
+        entity::{ActiveValue, ColumnTrait, EntityTrait},
+        query::{JsonValue, QueryFilter, QueryTrait},
+        sea_query::query::OnConflict,
+        ConnectionTrait, DbBackend, DbErr, TransactionTrait,
+    },
+    serde_json::value::Value,
+    solana_sdk::pubkey::Pubkey,
+    tokio::time::{sleep, Duration},
+    tracing::warn,
+};
+
+pub async fn burn_v1_asset<T: ConnectionTrait + TransactionTrait>(
+    conn: &T,
+    id: Pubkey,
+    slot: u64,
+) -> ProgramTransformerResult<()> {
+    let slot_i = slot as i64;
+    let model = asset::ActiveModel {
+        id: ActiveValue::Set(id.to_bytes().to_vec()),
+        slot_updated: ActiveValue::Set(Some(slot_i)),
+        burnt: ActiveValue::Set(true),
+        ..Default::default()
+    };
+    let mut query = asset::Entity::insert(model)
+        .on_conflict(
+            OnConflict::columns([asset::Column::Id])
+                .update_columns([asset::Column::SlotUpdated, asset::Column::Burnt])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+    query.sql = format!(
+        "{} WHERE excluded.slot_updated > asset.slot_updated",
+        query.sql
+    );
+    conn.execute(query).await?;
+    Ok(())
+}
+
+const RETRY_INTERVALS: &[u64] = &[0, 5, 10];
+
+pub async fn save_v1_asset<T: ConnectionTrait + TransactionTrait>(
+    conn: &T,
+    id: Pubkey,
+    account_data: &MplCoreAccountData,
+    slot: u64,
+) -> ProgramTransformerResult<Option<DownloadMetadataInfo>> {
+    // Notes:
+    // The address of the Core asset is used for Core Asset ID.  There are no token or mint accounts.
+    // There are no `MasterEdition` or `Edition` accounts associated with Core assets.
+    let id_array = id.to_bytes();
+    let id_vec = id_array.to_vec();
+
+    // Note: This indexes both Core Assets and Core Collections.
+    let asset = match account_data {
+        MplCoreAccountData::Asset(indexable_asset) => indexable_asset,
+        MplCoreAccountData::Collection(indexable_asset) => indexable_asset,
+        _ => return Err(ProgramTransformerError::NotImplemented),
+    };
+
+    //-----------------------
+    // Asset authority table
+    //-----------------------
+
+    // If it is an `Address` type, use the value directly.  If it is a `Collection`, search for and
+    // use the collection's authority.
+    let update_authority = match asset.update_authority {
+        UpdateAuthority::Address(address) => address.to_bytes().to_vec(),
+        UpdateAuthority::Collection(address) => {
+            find_collection_authority(conn, address.to_bytes().to_vec(), RETRY_INTERVALS)
```

Comment:
yes it could slow down indexing, but honestly not sure how to deal with it other than a retry.  This approach is similar to what token metadata indexing uses to look for the token.

Reply:
In real life, for mpl-core the collection will have to exist before the assets are minted to the collection or if the asset is added to the collection, so this should not fail as long as concurrent process indexing got the collection.